### PR TITLE
perf(autocomplete): Use virtual repeat...

### PR DESCRIPTION
**Pull request by kseamon**
...and some other optimizations

### CODE REVIEW

**Comment by kseamon** [comment]
I gathered some numbers (All in ms) from the chrome profiler testing dist/demos/autocomplete/demoBasicUsage/index.html, which has 50 repeated items. I'd expect that with a larger data set (say hundreds of items or more), the times would diverge by a lot more.

Startup times:
Old version: 121.6, 109.0, 104.9
With virtual repeat: 112.1, 105.6, 115.2

ctrl.matches (which feeds the repeater) is not populated yet, so we shouldn't expect much change until the user starts typing...

Time to click into textfield and type "aaa":
Old version: 122.6, 124.5, 122.3
With virtual repeat: 46.5, 46.4, 45.0

**Comment by jelbourn** [comment]
@robertmesserle

File: src/components/autocomplete/autocomplete.scss [hunk 1]
```diff
@@ -178,14 +178,18 @@ md-autocomplete {
     }
   }
 }
+
+.md-autocomplete-suggestions-container {
+  position: absolute !important;
```
Comment (robertmesserle) [inline src/components/autocomplete/autocomplete.scss]: Can we do this without using `!important`?

File: src/components/autocomplete/js/autocompleteDirective.js [hunk 2]
```diff
@@ -165,19 +170,17 @@ function MdAutocomplete ($mdTheming, $mdUtil) {
...
-        var templateTag = element.find('md-item-template').remove(),
+        var templateTag = element.find('md-item-template').detach(),
             html = templateTag.length ? templateTag.html() : element.html();
         if (!templateTag.length) element.empty();
         return html;
       }
 
       function getNoItemsTemplate() {
-        var templateTag = element.find('md-not-found').remove(),
+        var templateTag = element.find('md-not-found').detach(),
```
Comment (robertmesserle) [inline src/components/autocomplete/js/autocompleteDirective.js]: I believe this is just being used as a template (and angular isn't actually parsing it) - does it matter if we use remove or detach?

**Comment by robertmesserle** [comment]
LGTM

File: src/components/autocomplete/autocomplete.spec.js [hunk 3]
```diff
@@ -27,6 +27,14 @@ describe('<md-autocomplete>', function() {
     return scope;
   }
 
+  function waitForVirtualRepeat(element) {
+    inject(function ($$rAF) {
+      $$rAF.flush();
+      element.scope().$apply();
+      $$rAF.flush();
+    });
```
Comment (gkalpak) [inline src/components/autocomplete/autocomplete.spec.js]: So, I think I get it: You just have to `scope.$apply()`, then, `$timeout.flush()`, then `$$rAF.flush()`, then again `scope.$apply()`, then finally `$$rAF.flush()` again and...BAM...virtual-repeat ready !!!
Nice !

<sub>(Some comments would be also nice here, but I guess that's not very straightforward.)</sub>

File: src/components/autocomplete/js/highlightController.js [hunk 4]
```diff
@@ -2,26 +2,33 @@ angular
...
 
-  return init();
+  function init (termExpr, unsafeTextExpr) {
+    var text = null,
+        regex = null,
+        flags = $attrs.mdHighlightFlags || '',
+        watcher = $scope.$watch(function($scope) {
+          return {
+            term: termExpr($scope),
+            unsafeText: unsafeTextExpr($scope)
```
Comment (gkalpak) [inline src/components/autocomplete/js/highlightController.js]: Is `unsafeText` really changing ?
Is this a new feature ?

File: src/components/autocomplete/js/highlightDirective.js [hunk 5]
```diff
@@ -31,13 +31,17 @@ angular
...
 
-function MdHighlight () {
+function MdHighlight ($interpolate, $parse) {
   return {
     terminal: true,
-    scope: false,
     controller: 'MdHighlightCtrl',
-    link: function (scope, element, attr, ctrl) {
-      ctrl.init(attr.mdHighlightText);
+    compile: function mdHighlightCompile(tElement, tAttr) {
```
Comment (gkalpak) [inline src/components/autocomplete/js/highlightDirective.js]: Why use a `compile` function, when `link` works just as well ?

File: src/components/autocomplete/autocomplete.spec.js (hunk 3 above)
Comment (ThomasBurleson) [inline src/components/autocomplete/autocomplete.spec.js]: LOL.

File: src/components/autocomplete/js/autocompleteDirective.js (hunk 2 above)
Comment (kseamon) [inline src/components/autocomplete/js/autocompleteDirective.js]: https://github.com/angular/angular.js/blob/master/src/jqLite.js

remove will call into jqLiteDealloc, which tries to find data associated with the elements to clean up. In this case we can be confident that such data does not exist, so detach is faster.

File: src/components/autocomplete/js/highlightController.js (hunk 4 above)
Comment (kseamon) [inline src/components/autocomplete/js/highlightController.js]: Previously, it depended on the fact that ng-repeat would destroy and create a new instance of the directive if the underlying data changed. Virtual repeat reuses dom nodes (and thus the directives they contain), so they need to update on scope change.

File: src/components/autocomplete/js/highlightDirective.js (hunk 5 above)
Comment (kseamon) [inline src/components/autocomplete/js/highlightDirective.js]: Performance. The compile function is called only once by the repeater (link is called for each cloned item). This makes less of a difference now that we're not using ng-repeat, but it's still there.

File: src/components/autocomplete/js/autocompleteDirective.js (hunk 2 above)
Comment (robertmesserle) [inline src/components/autocomplete/js/autocompleteDirective.js]: Ah, very cool!

File: src/components/autocomplete/autocomplete.spec.js (hunk 3 above)
Comment (kseamon) [inline src/components/autocomplete/autocomplete.spec.js]: as @gkalpak says, the code is 100% self-documenting and not confusing at all.

However, I've added some comments as well.

**Comment by ThomasBurleson** [comment]
@kseamon - please rebase and I can merge.

**Comment by kseamon** [comment]
Rebased

**Comment by robertmesserle** [comment]
@jelbourn This is failing due to tests relating to Calendar, can you take a look and help me merge this in tomorrow?

**Comment by jelbourn** [comment]
It's probably because calendar tests need to do the same sequence of flushes / applies.

**Comment by ThomasBurleson** [comment]
@robertmesserle, @jelbourn - tests still failing:

```console
Firefox 31.0.0 (Mac OS X 10.9.0) md-calendar calendar construction should highlight the selected date FAILED
	Error: [$compile:ctreq] Controller 'mdCalendar', required by directive 'mdCalendarMonth', can't be found!
	http://errors.angularjs.org/1.4.4/$compile/ctreq?p0=mdCalendar&p1=mdCalendarMonth in /Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js (line 68)
	minErr/<@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:68:5
	getControllers@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:8105:1
	getControllers@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:8110:13
	nodeLinkFn@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:8250:1
	compositeLinkFn@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:7638:1
	publicLinkFn@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:7512:30
	createBoundTranscludeFn/boundTranscludeFn@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:7659:11
	controllersBoundTransclude@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:8273:11
	VirtualRepeatController.prototype.getBlock_@/Users/thomasburleson/Documents/projects/Google/projects/material/src/components/virtualRepeat/virtualRepeater.js:630:3
	VirtualRepeatController.prototype.virtualRepeatUpdate_@/Users/thomasburleson/Documents/projects/Google/projects/material/src/components/virtualRepeat/virtualRepeater.js:581:5
	bind/<@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:1185:1
	$watchCollectionAction@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:15618:13
	$RootScopeProvider/this.$get</Scope.prototype.$digest@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:15753:23
	$RootScopeProvider/this.$get</Scope.prototype.$apply@/Users/thomasburleson/Documents/projects/Google/projects/material/node_modules/angular/angular.js:16024:13
	applyDateChange@/Users/thomasburleson/Documents/projects/Google/projects/material/src/components/datepicker/calendar.spec.js:19:5
	@/Users/thomasburleson/Documents/projects/Google/projects/material/src/components/datepicker/calendar.spec.js:243:7
```

**Comment by kseamon** [comment]
@ThomasBurleson @jelbourn Updated - Tests are fixed

**Comment by lmadeira** [comment]
Autocomplete with custom template is not work properly after this commit even at: https://material.angularjs.org/HEAD/#/demo/material.components.autocomplete
